File: C# Examples/13 Document Manager API/SWDMProperties/MyDocMan.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using SolidWorks.Interop.swdocumentmgr;
using System.Windows.Forms;
using System.IO;
using System.Diagnostics;

namespace SWDMProperties
{
class MyDocMan
{
    SwDMApplication swDMApp;
    public SwDMApplication DocManApp
    {   
        get {
            if (swDMApp == null)
            {
                SwDMClassFactory swDMClassFact = new SwDMClassFactory();
                swDMApp = (SwDMApplication)swDMClassFact.GetApplication("ENTER LICENSE KEY HERE");                    
            }
            return swDMApp; }
    }

        #region File Operations
        public string BrowseForDoc()
        {
            string filePath = "";
            OpenFileDialog ofd = new OpenFileDialog();
            ofd.Filter = "SOLIDWORKS Files (*.sld*)|*.sld*"
                + "|SOLIDWORKS Part (*.sldprt)|*.sldprt"
                + "|SOLIDWORKS Assembly (*.sldasm)|*.sldasm"
                + "|SOLIDWORKS Drawing (*.slddrw)|*.slddrw"
                + "|All files (*.*)|*.*";
            if (ofd.ShowDialog() == DialogResult.OK)
            {
                filePath = ofd.FileName;
            }
            if (filePath == "")
            {
                return null;
            }
            return filePath;
        }

        public SwDMDocument OpenDoc(string filePath, bool isReadOnly)
        {
            string ext = Path.GetExtension(filePath).ToUpper();
            SwDmDocumentType docType;
            switch (ext)
            {
                case "SLDPRT":
                    docType = SwDmDocumentType.swDmDocumentPart;
                    break;
                case "SLDASM":
                    docType = SwDmDocumentType.swDmDocumentAssembly;
                    break;
                case "SLDDRW":
                    docType = SwDmDocumentType.swDmDocumentDrawing;
                    break;
                default:
                    docType = SwDmDocumentType.swDmDocumentUnknown; 
                    break;
            }

            SwDmDocumentOpenError openerrors = 0;
            SwDMDocument dmDoc = DocManApp.GetDocument(filePath,
                docType, isReadOnly, out openerrors);

            //error handling
            if (dmDoc == null)
            {
                string addToMessage;
                switch (openerrors)
                {
                    case SwDmDocumentOpenError.swDmDocumentOpenErrorFail:
                        addToMessage = "Failed to access the document.";
                        break;
                    case SwDmDocumentOpenError.swDmDocumentOpenErrorNonSW:
                        addToMessage = "Not a SOLIDWORKS file.";
                        break;
                    case SwDmDocumentOpenError.swDmDocumentOpenErrorFileNotFound:
                        addToMessage = "File not found.";
                        break;
                    case SwDmDocumentOpenError.swDmDocumentOpenErrorFileReadOnly:
                        addToMessage = "File is Read Only or is open in another application.";
                        break;
                    case SwDmDocumentOpenError.swDmDocumentOpenErrorNoLicense:
                        addToMessage = "Invalid Document Manager license.";
                        break;
                    case SwDmDocumentOpenError.swDmDocumentOpenErrorFutureVersion:
                        addToMessage = "Unsupported file version.";
                        break;
                    default:
                        addToMessage = "";
                        break;
                }
                string errMessage = "Error accessing " + filePath 
                    + System.Environment.NewLine + addToMessage;
                MessageBox.Show(errMessage, "SWDMProperties", MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                return null;
            }
            return dmDoc;
        }
        #endregion


        #region Properties
        public string ReadAllProps(SwDMDocument dmDoc)
        {
            //read all file custom properties
            string output;
            output = dmDoc.FullName + " Properties:";
            object[] propNames = dmDoc.GetCustomPropertyNames();
            if (propNames != null)
            {
                foreach (string propName in propNames)
                {
                    string propVal;
                    SwDmCustomInfoType myType;
                    propVal = dmDoc.GetCustomProperty(propName, out myType);
                    output += System.Environment.NewLine + propName + "\t" + propVal;
                }
            }
            return output;
        }

        public string ReadConfiProps(SwDMDocument dmDoc)
        {
            //read all configurations
            ISwDMConfigurationMgr configMan;
            configMan = dmDoc.ConfigurationManager;
            try
            {
                object[] configNames = configMan.GetConfigurationNames();
                string output = "Configurations:";
                foreach (string config in configNames)
                {
                    output += "\nConfiguration: " + config;
                    //get configuration specific properties
                    SwDMConfiguration dmConfig;
                    dmConfig = configMan.GetConfigurationByName(config);
                    object[] propNames = dmConfig.GetCustomPropertyNames();
                    if (propNames != null)
                    {
                        foreach (string propName in propNames)
                        {
                            string propVal;
                            SwDmCustomInfoType myType;
                            propVal = dmConfig.GetCustomProperty(propName, out myType);
                            output += System.Environment.NewLine + propName + "\t" + propVal;
                        }
                    }
                }            
                return output; 
            }
            catch (Exception)
            {
                //ignore the error for drawings
                return "";                
            }
            
        }

        public void WriteProp(SwDMDocument dmDoc,
            string pName, string pVal)
        {
            //write a property
            bool addRes;
            addRes = dmDoc.AddCustomProperty(pName,
                SwDmCustomInfoType.swDmCustomInfoText, pVal);
            if (!addRes)
            {
                dmDoc.SetCustomProperty(pName, pVal);
            }
        }
        #endregion

        #region Tables
        public string GetBOMTableText(SwDMDocument10 dmDoc)
        {
            string tableText = "";
            object[] tableNames = dmDoc.GetTableNames(SwDmTableType.swDmTableTypeBOM);
            foreach (string table in tableNames)
            {
                SwDMTable5 dmTable = (SwDMTable5)dmDoc.GetTable(table);
                Debug.Print(dmTable.Name);
                int rows;
                int columns;
                SwDmTableError errors;
                string[] ttext = dmTable.GetTableCellText(out errors, out rows, out columns);
                tableText = "Table: " + dmTable.Name + Environment.NewLine;
                //format the cell text array into a tab delimited text file
                int cells = 0;
                for (int i = 1; i <= rows; i++)
                {
                    string rowText = "";
                    for (int j = 1; j <= columns; j++)
                    {
                        rowText += ttext[cells] + "\t";
                        cells++;
                    }
                    tableText += rowText + Environment.NewLine;
                }
            }
            return tableText;
        }
        #endregion

        #region References
        //recursive routine to report all missing file references
        public void GetRefs(string filePath, bool replace = false)
        {
            Debug.Print("=== " + filePath + " REFERENCES ===");
            //get references
            SwDMSearchOption searchOpt;
            object brokenRefs;
            object isVirtual;
            object timeStamp;
            object refs;
            searchOpt = DocManApp.GetSearchOptionObject();
            searchOpt.SearchFilters = (int)SwDmSearchFilters.SwDmSearchExternalReference;
            SwDMDocument21 dmDoc = (SwDMDocument21)OpenDoc(filePath, false);
            if (dmDoc == null)
            {
                return;
            }
            refs = dmDoc.GetAllExternalReferences5(searchOpt, out brokenRefs,
                out isVirtual, out timeStamp, out refs);
            string myRef;
            string[] references = (string[])refs;
            for (int i = 0; i < references.Length; i++)
            {
                myRef = references[i];
                if (!(bool)isVirtual)
                {
                    if (!File.Exists(myRef))
                    {
                        Debug.Print("MISSING: " + myRef);
                        if (replace)
                        {
                            DialogResult msgRes;
                            msgRes = MessageBox.Show(text: "Missing: \n"
                                + myRef + "\nWould you like to replace it?",
                                caption: "SWDMReferences",
                                buttons: MessageBoxButtons.YesNo, 
                                icon: MessageBoxIcon.Question);
                            if (msgRes == DialogResult.Yes)
                            {
                                string newFilePath;
                                newFilePath = BrowseForDoc(Path.GetFileName(myRef));
                                if (newFilePath != "")
                                {
                                    dmDoc.ReplaceReference(myRef, newFilePath);
                                }
                            }
                        }
                    }
                    else
                    {
                        Debug.Print(myRef);
                        string ext = Path.GetExtension(myRef);
                        if (ext.ToUpper() == ".SLDASM")
                        {
                            //recursively search the assembly
                            //for its references
                            GetRefs(myRef, replace);
                        }
                    }
                }
            }
            //save if replacing references
            if (replace){dmDoc.Save();}            
            dmDoc.CloseDoc();
        }

        public string BrowseForDoc(string fileName = "")
        {
            //open the document for read/write
            string filePath = "";
            OpenFileDialog ofd = new OpenFileDialog();
            ofd.Filter = "SOLIDWORKS Files (*.sldprt, "
              + "*.sldasm, *.slddrw)"
              + "|*.sldprt;*.sldasm;*.slddrw|Parts "
              + "(*.sldprt)|*.sldprt"
              + "|Assemblies (*.sldasm)|*.sldasm|Drawings "
              + "(*.slddrw)"
              + "|*.slddrw|All files (*.*)|*.*";
            ofd.FileName = fileName;
            if (ofd.ShowDialog() == DialogResult.OK)
            {
                filePath = ofd.FileName;
            }
            if (filePath == "")
            {
                return null;
            }
            return filePath;
        }
        #endregion
    }
}

File: 13 Document Manager API/SWDMProperties/MyDocMan.vb
Imports SolidWorks.Interop.swdocumentmgr
Imports System.Diagnostics

Public Class MyDocMan
  Dim swDMApp As SwDMApplication4

  Public ReadOnly Property DocManApp() As SwDMApplication4
    Get
      If swDMApp Is Nothing Then
        Dim swDMClassFact As SwDMClassFactory
        swDMClassFact =
          CreateObject("SwDocumentMgr.SwDMClassFactory")
        swDMApp = swDMClassFact.GetApplication("ENTER YOUR DOCUMENT MANAGER KEY HERE")
      End If
      Return swDMApp
    End Get
  End Property

  Friend Function BrowseForDoc(Optional ByVal FileName As String = "") As String
    'open the document for read/write
    Dim FilePath As String = ""
    Dim ofd As New Windows.Forms.OpenFileDialog
    ofd.Filter = "SOLIDWORKS Files (*.sldprt, " _
    & "*.sldasm, *.slddrw)" _
    & "|*.sldprt;*.sldasm;*.slddrw|Parts " _
    & "(*.sldprt)|*.sldprt" _
    & "|Assemblies (*.sldasm)|*.sldasm|Drawings " _
    & "(*.slddrw)" _
    & "|*.slddrw|All files (*.*)|*.*"
    ofd.FileName = FileName
    If ofd.ShowDialog = Windows.Forms.DialogResult.OK Then
      FilePath = ofd.FileName
    End If
    If FilePath = "" Then Return Nothing

    Return FilePath
  End Function

  Friend Function OpenDoc(FilePath As String,
   isReadOnly As Boolean) As SwDMDocument
    Dim ext As String = System.IO.Path.GetExtension(FilePath).ToUpper
    Dim docType As Integer
    Select Case ext
      Case "SLDPRT"
        docType = SwDmDocumentType.swDmDocumentPart
      Case "SLDASM"
        docType = SwDmDocumentType.swDmDocumentAssembly
      Case "SLDDRW"
        docType = SwDmDocumentType.swDmDocumentDrawing
      Case Else
        docType = SwDmDocumentType.swDmDocumentUnknown
    End Select

    Dim openerrors As Integer
    Dim dmDoc As SwDMDocument
    dmDoc = DocManApp.GetDocument(FilePath,
      docType, isReadOnly, openerrors)

    'error handling
    If dmDoc Is Nothing Then
      Dim errMessage As String = "Error accessing " & FilePath
      Select Case openerrors
        Case SwDmDocumentOpenError.swDmDocumentOpenErrorNonSW
                    errMessage += vbCrLf & "Non SOLIDWORKS file."
                Case SwDmDocumentOpenError.swDmDocumentOpenErrorNoLicense
                    errMessage += vbCrLf & "Invalid Document " _
            & "Manager license."
                Case SwDmDocumentOpenError.swDmDocumentOpenErrorFutureVersion
                    errMessage += vbCrLf & "Unsupported file version."
                Case SwDmDocumentOpenError.swDmDocumentOpenErrorFileNotFound
                    errMessage += vbCrLf & "File not found."
                Case SwDmDocumentOpenError.swDmDocumentOpenErrorFail
                    errMessage += vbCrLf & "Failed to access the document."
                Case SwDmDocumentOpenError.swDmDocumentOpenErrorFileReadOnly
                    errMessage += vbCrLf & "File is Read Only or is " _
            & "open in another application."
            End Select
      MsgBox(errMessage, MsgBoxStyle.Exclamation)
      Return Nothing
    Else
      Return dmDoc
    End If
  End Function

  Friend Function ReadAllProps(ByVal dmDoc As SwDMDocument) As String
    'read all file custom properties
    Dim output As String
    output = dmDoc.FullName & " Properties:"
    Dim propNames As Object = dmDoc.GetCustomPropertyNames()
    If Not propNames Is Nothing Then
      For Each propName As String In propNames
        Dim propVal As String
        propVal = dmDoc.GetCustomProperty(propName, Nothing)
        output = output & vbCrLf & propName & vbTab & propVal
      Next
    End If
    Return output
  End Function

  Friend Function ReadConfigProps(ByVal dmDoc As SwDMDocument) As String
    'read all configurations
    Dim configMan As SwDMConfigurationMgr
    configMan = dmDoc.ConfigurationManager()
    Dim configNames As Object
    Try
      configNames = configMan.GetConfigurationNames
      Dim output As String = "Configurations:"
      For Each config As String In configNames
        output = output & vbCrLf & "Configuration: " & config
        'Get configuration specific properties
        Dim dmConfig As SwDMConfiguration
        dmConfig = configMan.GetConfigurationByName(config)
        Dim propNames As Object = dmConfig.GetCustomPropertyNames()
        If Not propNames Is Nothing Then
          For Each propName As String In propNames
            Dim propVal As String
            propVal = dmConfig.GetCustomProperty(propName, Nothing)
            output = output & vbCrLf & propName & vbTab & propVal
          Next
        End If

      Next
      Return output
    Catch ex As Exception
      'ignore the error for drawings
      Return ""
    End Try

  End Function

  Friend Sub WriteProp(ByVal dmDoc As SwDMDocument,
  ByVal pName As String, ByVal pVal As String)
    'write a property
    Dim addRes As Boolean
    addRes = dmDoc.AddCustomProperty(pName,
      SwDmCustomInfoType.swDmCustomInfoText, pVal)
    If Not addRes Then
      dmDoc.SetCustomProperty(pName, pVal)
    End If
  End Sub

  Friend Function GetBOMTableText(dmDoc As SwDMDocument10) As String
    Dim tableText As String
    Dim tableNames As Object
    tableNames = dmDoc.GetTableNames(SwDmTableType.swDmTableTypeBOM)
    For Each table As String In tableNames
      Dim dmTable As SwDMTable5
      dmTable = dmDoc.GetTable(table)
      Debug.Print(dmTable.Name)
      Dim rows As Integer
      Dim columns As Integer
      Dim ttext As Object
      ttext = dmTable.GetTableCellText(Nothing, rows, columns)

      tableText = "Table: " & dmTable.Name & vbCrLf
      'format the cell text array into a tab 
      'delimited text file
      Dim cells As Integer = 0
      For i As Integer = 1 To rows
        Dim rowText As String = ""
        For j As Integer = 1 To columns
          rowText = rowText & ttext(cells) & vbTab
          cells += 1
        Next
        tableText = tableText & rowText & vbCrLf
      Next
    Next
    Return tableText
  End Function














  Friend Function GetBOMTableTextFULL(dmDoc As SwDMDocument10) As String
    Dim tableText As String
    Dim tableNames As Object
    tableNames = dmDoc.GetTableNames(SwDmTableType.swDmTableTypeBOM)
    For Each table As String In tableNames
      Dim dmTable As SwDMTable5
      dmTable = dmDoc.GetTable(table)
      Debug.Print(dmTable.Name)
      Dim rows As Integer
      Dim columns As Integer
      Dim ttext As Object
      ttext = dmTable.GetTableCellText(Nothing, rows, columns)
      tableText = "Table: " & dmTable.Name & vbCrLf
      'format the cell text array into a tab 
      'delimited text file
      Dim cells As Integer = 0
      For i As Integer = 1 To rows
        Dim rowText As String = ""
        For j As Integer = 1 To columns
          rowText = rowText & ttext(cells) & vbTab
          cells += 1
        Next
        tableText = tableText & rowText & vbCrLf
      Next
    Next
    Return tableText
  End Function

  'Recursive routine to report all missing file references
  Public Sub GetRefs(FilePath As String,
Optional replace As Boolean = False)
    Debug.Print("=== " & FilePath & " REFERENCES ===")
    'get references
    Dim searchOpt As SwDMSearchOption
    Dim brokenRefs As Object = Nothing
    Dim isVirtual As Object = Nothing
    Dim timeStamp As Object = Nothing
    Dim refs As Object
    searchOpt = DocManApp.GetSearchOptionObject()

    searchOpt.SearchFilters =
    SwDmSearchFilters.SwDmSearchExternalReference

    Dim dmDoc As SwDMDocument21
    dmDoc = OpenDoc(FilePath, False)
    If dmDoc Is Nothing Then Exit Sub

    refs = dmDoc.GetAllExternalReferences5(
    searchOpt, brokenRefs, isVirtual, timeStamp, Nothing)

    Dim ref As String
    For i As Integer = 0 To refs.Length - 1
      ref = refs(i)
      If Not isVirtual(i) Then
        If Not IO.File.Exists(ref) Then
          Debug.Print("MISSING: " & ref)
          If replace Then
            Dim msgRes As MsgBoxResult
            msgRes = MsgBox("Missing: " & vbCrLf & ref & vbCrLf _
            & "Would you like to replace it?",
            MsgBoxStyle.YesNo + MsgBoxStyle.Question)
            If msgRes = MsgBoxResult.Yes Then
              Dim newFilePath As String
              newFilePath = BrowseForDoc(IO.Path.GetFileName(ref))
              If newFilePath <> "" Then
                dmDoc.ReplaceReference(ref, newFilePath)
              End If
            End If
          End If

        Else
          Debug.Print(ref)
          If IO.Path.GetExtension(ref).ToUpper = ".SLDASM" Then
            'recursively search the assembly 
            'for its references
            GetRefs(ref, replace)
          End If
        End If
      End If
    Next
    'save if replacing references
    If replace Then dmDoc.Save()
    dmDoc.CloseDoc()
  End Sub


End Class

